core/src/main/java/hudson/security/HudsonPrivateSecurityRealm.java

```diff
@@ -925,51 +933,195 @@ public boolean isHashValid(String hash) {
+        private static byte[] generateSalt() {
+            SecureRandom random = new SecureRandom();
+            byte[] salt = new byte[16];
+            random.nextBytes(salt);
+            return salt;
+        }
+
+        @SuppressFBWarnings(value = "MS_SHOULD_BE_FINAL", justification = "Accessible via System Groovy Scripts")
+        @Restricted(NoExternalUse.class)
+        private static int MAXIMUM_PBKDF2_LOG_ROUND = SystemProperties.getInteger(HudsonPrivateSecurityRealm.class.getName() + ".maximumPBKDF2LogRound", (int) Math.pow(2, 10));
+
+        private static final Pattern PBKDF2_PATTERN = Pattern.compile("^\\$PBKDF2\\$HMACSHA512\\:([0-9]{4})\\:[a-zA-Z0-9=]+\\$[a-zA-Z0-9=]+");
+
+        @Override
+        public boolean isHashValid(String hash)    {
+            Matcher matcher = PBKDF2_PATTERN.matcher(hash);
+            if (matcher.matches()) {
+                String logNumOfRound = matcher.group(1);
+                // no number format exception due to the expression
+                int logNumOfRoundInt = Integer.parseInt(logNumOfRound);
+                if (logNumOfRoundInt > 0 && logNumOfRoundInt <= MAXIMUM_PBKDF2_LOG_ROUND) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+
+        private static boolean validatePassword(String originalPassword, String storedPassword) {
+            String[] parts = storedPassword.split("[:$]");
+            int iterations = Integer.parseInt(parts[3]);
+
+            byte[] salt = fromHex(parts[4]);
+            byte[] hash = fromHex(parts[5]);
+
+            PBEKeySpec spec = new PBEKeySpec(originalPassword.toCharArray(),
+                    salt, iterations, hash.length * HASH_LENGTH);
+
+            byte[] generatedHashValue = generateSecretKey(spec);
+
+            int diff = hash.length ^ generatedHashValue.length;
+            for (int i = ZERO; i < hash.length && i < generatedHashValue.length; i++)
+            {
+                diff |= hash[i] ^ generatedHashValue[i];
+            }
+            return diff == 0;
+        }
+
+        private static byte[] fromHex(String storedPassword)  {
+            byte[] bytes = new byte[storedPassword.length() / HEX_TWO];
+            for (int i = ZERO; i < bytes.length; i++)
+            {
+                bytes[i] = (byte) Integer.parseInt(storedPassword.substring(HEX_TWO * i, HEX_TWO * i + HEX_TWO), RADIX_LENGTH);
+            }
+            return bytes;
+        }
+
+    }
+
+
+
+    /* package */ static final PasswordHashEncoder PASSWORD_HASH_ENCODER = getPasswordHashEncoder();
+
+    public static PasswordHashEncoder getPasswordHashEncoder() {
+        return Util.FIPS_MODE ? new PBKDF2PasswordEncoder() : new JBCryptEncoder();
+    }
+
+
+    public static final String PBKDF2 = "#pbkdf2:";
```

Comment:
Prefix for PBKDF2 seems to be redundant since the new format starts with: $PBKDF2. So it shall be used where there is check using it as a $PBKDF2 - but there is no need to concatenate/multiply this or similar in the final string.